tools/src/h5diff/ph5diff_main.c

```diff
@@ -62,7 +64,16 @@ main(int argc, const char *argv[])
     outBuffOffset = 0;
     g_Parallel    = 1;
 
-    MPI_Init(&argc, (char ***)&argv);
+    if ((args = (char **)HDcalloc((size_t)argc, sizeof(char *))) == NULL) {
```

Comment:
There's no need to copy the args to fix this warning. The signature for main should simply take a char * array rather than a const char * array. Just dropping the const in the signature should be enough.

Reply:
Then, ```parse_command_line(int argc, const char *argv[], const char **fname1, const char **fname2,``` will trigger warnings.

Reply:
In that case, things have to be fixed further down the stack as well. main should be changed here to simply take a char * array, which will fix the warning from MPI_Init. parse_command_line will probably need some changing as well.

Reply:
I disagree. MPI_Init should use const. Until then, we can add another layer like HDMPI_Init() since many HDF5 CLI tools use const in main() function.

Reply:
It is a mistake for the tools to use const for argv. This is not the prototype for main and it's very clearly the wrong thing to do because we act like argv is const and then immediately cast that away when passing it to MPI_Init. Note that MPI_Init might modify argc/argv so it should NOT accept const parameters. _All_ of HDF5's tools should be updated to not use a const qualifier for argv and it was likely done just to paper over the warnings caused by trying to pass argv directly to H5_get_option which takes a const argv.

Reply:
From https://www.open-mpi.org/doc/current/man3/MPI_Init.3.php
```
All MPI programs must contain a call to MPI_Init or MPI_Init_thread. Open MPI accepts the C/C++ argc and argv arguments to main, but neither modifies, interprets, nor distributes them:
```
I don't know who (@qkoziol  or @derobins  ?) started adding the const in argv[] for main() and H5_get_option() in src/H5systemc. but I think it's the right thing to do for many reasons.

Reply:
See also: https://stackoverflow.com/questions/9185792/passing-arguments-via-command-line-with-mpi.

MPI will not consume what the caller has passed to the program itself but may consume any arguments to mpirun so that it effectively looks like those arguments were never passed in the first place. I think it was in good spirit to add const to H5_get_option, but it is definitely NOT the right thing to do for main. Regardless of whether the MPI implementation modifies argc/argv or not, MPI_Init's prototype does not use const for its arguments and simply casting the const away from the const qualified argv isn't good. If the tools really want to use a const qualified argv, then they should either copy the arguments into a new array or just cast argv appropriately. But NONE of the tools should use a const argv, since this doesn't match main's prototype as given by the C standard.

Reply:
The C standard is for legacy when **const** was not available in compiler. Why do you think the modern compiler Objective-C / clang added **const**?

Reply:
What does C99 standard indicate -  we have moved to that minimum for develop.

Reply:
> What does C99 standard indicate - we have moved to that minimum for develop.

The C standard only mandates that the implementation should require either void for main or the `int argc, char * argv[]` combo. In practice, pretty much all implementations will allow const char * argv[], but that doesn't mean we should be using that. Clearly this is a case where using a const argv is a mistake.

Reply:
Agreed.

Reply:
The C standard requires either int main(void) or int main(int argc, char* argv[]). Equivalents are allowed, like char **argv, but the const should be removed.

Reply:
Also: "The parameters argc and argv and the strings pointed to by the argv array shall be modifiable by the program, and retain their last-stored values between program startup and program termination."